docs/rfcs/parcel-v3/0001_cache_implementation.md

```diff
@@ -0,0 +1,283 @@
+
+We will segment the caches between the current JavaScript implementation and the new Rust implementation.
+
+For all Rust back-ed requests, we will disable caching and allow Rust to manage reading from its own cache
+implementation.
+
+## Serialization format
+
+We will research and benchmark a few options for serialization, then aim to make a decision trading-off serialization
+performance, stability and size.
+
+## Large blob entries
+
+We will deprecate hybrid file-system and LMDB caches. All entries will be written to LMDB.
+
+## Remote caches
+
+We should consider that at some point the cache contents may be segmented between a local and a remote cache. That is,
+a cache server URL might be provided and the bundler should be able to fetch from it if a cache entry is missing
+locally.
+
+## LMDB wrapper library
+
+Suggest we use:
+
+- https://github.com/meilisearch/heed
+
+Which seems mature enough due to usage in `meilisearch`.
+
+### LMDB tuning
+
+We should flip the following [LMDB flags](http://www.lmdb.tech/doc/group__mdb__env.html#ga5791dd1adb09123f82dd1f331209e12e):
+
+- `NOSYNC` - Do not call [`fsync`](https://man7.org/linux/man-pages/man2/fsync.2.html) after commit. This means
+  potentially our cache won't be durable (if the system shuts-down before the OS is able to flush data, it could be
+  there would be data loss, even after writing), but we don't care about durability for our use-case.
+- `MAPASYNC` - Do not call [`msync`](https://man7.org/linux/man-pages/man2/msync.2.html) after writing. This is
+  the same effect.
+- `NOMETASYNC` - Do not `fsync` the metapage after commit
+
+This greatly improves write performance.
+
+### LMDB and RKYV
+
+Unfortunately `rkyv` and LMDB can't be used together. You can see a failing test case reproduction on commit
+`d75dda3af93179c58a4b55fdcc42a6a60e7f3e2f`. RKYV code was available on the `lmdb-cache-implementation` branch until
+being removed on commit `71e7fc269881a44bab0d44b054e53ce8beabc477`.
+
+`rkyv` requires values to be aligned by 4-byte words - https://rkyv.org/architecture/alignment.html. Since LMDB has no
+alignment guarantee on values, in order to use `rkyv` we'd be forced to copy the LMDB buffers onto an aligned location.
+
+This breaks the point of using zero-copy serialization, since in order to use `rkyv` with LMDB we'll need to copy the
+buffers into aligned memory locations or to try to get LMDB to align its internal structures.
+
+The issue is discussed on https://github.com/AltSysrq/lmdb-zero/issues/8 and https://github.com/meilisearch/heed/issues/198.
+
+### Orders of magnitude for LMDB write performance from rust
+
+This has been tested on an EC2 `c7i.8xlarge` instance using criterion.
+
+- Serializing a single request - **165ns** with `bincode` and 288ns with `rkyv`
+- Read a single request and deserialize - **821ns**
+- Write 1 request with async writes - **15204ns** / **15us**
+- Write 1 request with sync writes - **3000000ns** / **3000us** / **3ms**
```

Comment:
That's expensive! That means if you have 100,000 nodes and 500,000 edges in the RequestGraph (not unreasonable, my current benchmark has 87k nodes and 400k edges, and I'm sure Jira has 10x more than that), just writing them alone will take 30 minutes.

Even with async writes, that would be 9 seconds added to the build time just to write request nodes to the cache. That's 11x slower than performing a complete build from scratch with no cache.

Reply:
Note we'd write only request nodes so those estimates are 6x higher than reality.

It's quite expensive though, I think there is likely still some tuning that could speed it up a lot ; or something my benchmark does incorrectly.

Reply:
Let's iterate on and discuss this in the next few weeks as we get closer to having an E2E